rclpy/src/rclpy/_rclpy_action.c

```diff
@@ -0,0 +1,885 @@
+    return NULL;
+  }
+
+  rcl_wait_set_t * wait_set = (rcl_wait_set_t *)PyCapsule_GetPointer(pywait_set, "rcl_wait_set_t");
+  if (!wait_set) {
+    return NULL;
+  }
+
+  if (PyCapsule_IsValid(pyentity, "rcl_action_client_t")) {
+    rcl_action_client_t * action_client =
+      (rcl_action_client_t *)PyCapsule_GetPointer(pyentity, "rcl_action_client_t");
+    bool is_feedback_ready = false;
+    bool is_status_ready = false;
+    bool is_goal_response_ready = false;
+    bool is_cancel_response_ready = false;
+    bool is_result_response_ready = false;
+    rcl_ret_t ret = rcl_action_client_wait_set_get_entities_ready(
+      wait_set,
+      action_client,
+      &is_feedback_ready,
+      &is_status_ready,
+      &is_goal_response_ready,
+      &is_cancel_response_ready,
+      &is_result_response_ready);
+    if (RCL_RET_OK != ret) {
+      PyErr_Format(
+        PyExc_RuntimeError,
+        "Failed to get number of ready entities for action client: %s",
+        rcl_get_error_string().str);
+      rcl_reset_error();
+      return NULL;
+    }
+
+    PyObject * result_list = PyList_New(5);
+    PyList_SET_ITEM(result_list, 0, PyBool_FromLong(is_feedback_ready));
+    PyList_SET_ITEM(result_list, 1, PyBool_FromLong(is_status_ready));
+    PyList_SET_ITEM(result_list, 2, PyBool_FromLong(is_goal_response_ready));
+    PyList_SET_ITEM(result_list, 3, PyBool_FromLong(is_cancel_response_ready));
+    PyList_SET_ITEM(result_list, 4, PyBool_FromLong(is_result_response_ready));
+    return result_list;
+  } else if (PyCapsule_IsValid(pyentity, "rcl_action_server_t")) {
+    rcl_action_server_t * action_server =
+      (rcl_action_server_t *)PyCapsule_GetPointer(pyentity, "rcl_action_server_t");
+    bool is_goal_request_ready = false;
+    bool is_cancel_request_ready = false;
+    bool is_result_request_ready = false;
+    bool is_goal_expired = false;
+    rcl_ret_t ret = rcl_action_server_wait_set_get_entities_ready(
+      wait_set,
+      action_server,
+      &is_goal_request_ready,
+      &is_cancel_request_ready,
+      &is_result_request_ready,
+      &is_goal_expired);
+    if (RCL_RET_OK != ret) {
+      PyErr_Format(
+        PyExc_RuntimeError,
+        "Failed to get number of ready entities for action server: %s",
+        rcl_get_error_string().str);
+      rcl_reset_error();
+      return NULL;
+    }
+
+    PyObject * result_list = PyList_New(4);
+    PyList_SET_ITEM(result_list, 0, PyBool_FromLong(is_goal_request_ready));
+    PyList_SET_ITEM(result_list, 1, PyBool_FromLong(is_cancel_request_ready));
+    PyList_SET_ITEM(result_list, 2, PyBool_FromLong(is_result_request_ready));
+    PyList_SET_ITEM(result_list, 3, PyBool_FromLong(is_goal_expired));
+    return result_list;
+  } else {
+    PyErr_Format(PyExc_RuntimeError, "'%s' is not a known entity", PyCapsule_GetName(pyentity));
+    return NULL;
+  }
+}
+
+#define OPTIONS_COPY_QOS_PROFILE(Options, Profile) \
+  if (PyCapsule_IsValid(py ## Profile, "rmw_qos_profile_t")) { \
```

Comment:
Wow, in doing so I think we've uncovered a bug. It would appear that a capsule for `rmw_qos_profile_t` is never created. Instead `rclpy_convert_to_py_qos_policy()` created a regular Python object of type `QosProfile`. The C struct is not stored (or reconstructed) anywhere.

https://github.com/ros2/rclpy/blob/523c403c60c12dc304564f2e681bf1d6dd637b9d/rclpy/src/rclpy/impl/common.h#L27-L47

Correct me if I'm wrong, but I think this means that all of the checks `_rclpy.c` are failing silently and QoS settings are never copied to the `rcl` layer:

https://github.com/ros2/rclpy/blob/0fafaec6b3764654413d0ce37839d7c7178245af/rclpy/src/rclpy/_rclpy.c#L1179

https://github.com/ros2/rclpy/blob/0fafaec6b3764654413d0ce37839d7c7178245af/rclpy/src/rclpy/_rclpy.c#L1689

https://github.com/ros2/rclpy/blob/0fafaec6b3764654413d0ce37839d7c7178245af/rclpy/src/rclpy/_rclpy.c#L1779

https://github.com/ros2/rclpy/blob/0fafaec6b3764654413d0ce37839d7c7178245af/rclpy/src/rclpy/_rclpy.c#L1936

Reply:
@clalancette Is there an issue opened somewhere? If not, I can create one.

Reply:
No, I hadn't opened one yet.  Thanks for looking at this!

Reply:
False alarm. I missed the call to [`rclpy_convert_from_py_qos_policy()`](https://github.com/ros2/rclpy/blob/0fafaec6b3764654413d0ce37839d7c7178245af/rclpy/src/rclpy/_rclpy.c#L3401) in `qos.py`:

https://github.com/ros2/rclpy/blob/0fafaec6b3764654413d0ce37839d7c7178245af/rclpy/rclpy/qos.py#L123-L130

@clalancette This is not the bug you're looking for :robot: